GlutenShuffleManager router cache throws on concurrent first-touch and on unregisterShuffle for uncached shuffle ids

### Backend

VL (Velox) — the logic is backend-agnostic and lives in `gluten-core`.

### Bug description

`ShuffleManagerRouter`'s inner `Cache` assumes a single-coordinator lifecycle: each shuffleId is `store`d exactly once, before any `get`/`remove`. That holds on the driver (`registerShuffle` is called single-threaded by the DAGScheduler), but **not** on executors, where the per-executor router cache is populated lazily by task threads via `getReader`/`getWriter` → `ensureShuffleManagerRegistered`. Two observable problems follow.

**1. Concurrent first-touch race (primary).**
On a multi-core executor, N tasks of the same *new* shuffleId call `getReader`/`getWriter` concurrently. In `ensureShuffleManagerRegistered`, the `has()` check and the `store()` call are not atomic, and `store()` uses `cache.compute` with `assert(m == null)`. `ConcurrentHashMap.compute` holds a per-bin lock, which makes the failure **deterministic for the losers**: exactly one thread installs the manager, and every other thread's remapping function observes `m != null` → the `assert` fires → `AssertionError` propagates out of `getReader`/`getWriter` → those tasks are marked FAILED. It is retry-recoverable (by retry time the cache is populated), but it produces intermittent, timing-dependent spurious task failures that count toward `spark.task.maxFailures` and add latency; a wide new-shuffle stage produces a burst.

**2. `unregisterShuffle` asserts for an uncached shuffleId.**
Spark broadcasts `RemoveShuffle` to **every** executor (`BlockManagerMasterEndpoint.removeShuffle` → each executor's `BlockManagerStorageEndpoint`). An executor that never participated in a given shuffle has no entry in its router cache, so `Cache.remove`'s `assert(manager != null)` fires. The `AssertionError` is caught by `doAsync`'s Future (no crash, no task/job failure), but every shuffle cleanup emits an ERROR-level stack trace on each non-participating executor plus a WARN on the driver, and `unregisterShuffle` returns a failed RPC instead of a boolean — polluting error logs/metrics and potentially masking genuine removal errors.

Note: Scala's `Predef.assert` is an ordinary method call that is **not** gated by the JVM `-ea` flag, and the Gluten build sets no `-Xelide-below`, so these assertions fire in both test and shipped builds.

`GlutenShuffleManager` is currently annotated `@Experimental`.

A third manifestation of the same root cause is latent: `Cache.get` (reached via `BlockResolver`) also asserts, but every path that reaches it is on an already-registered executor, so it is not a reachable crash on a supported configuration and is out of scope here.

### Gluten version

main (1.8.0-SNAPSHOT)

### Spark version

Version-agnostic (applies to spark-3.3 / 3.4 / 3.5 / 4.0 / 4.1).

### Spark configurations

- `spark.shuffle.manager=org.apache.spark.shuffle.GlutenShuffleManager`
- Problem 1 requires `spark.executor.cores > 1` (multi-core executor).
- Problem 2 requires a multi-executor cluster.

### System information

N/A — logic issue in `gluten-core/src/main/scala/org/apache/spark/shuffle/ShuffleManagerRouter.scala`, independent of OS/hardware.

### Relevant logs

```text
java.lang.AssertionError: assertion failed: Shuffle manager was already cached for shuffle id: <id>
  at org.apache.spark.shuffle.ShuffleManagerRouter$Cache.store(ShuffleManagerRouter.scala:109)
  at org.apache.spark.shuffle.ShuffleManagerRouter.ensureShuffleManagerRegistered(ShuffleManagerRouter.scala:95)
  at org.apache.spark.shuffle.ShuffleManagerRouter.getReader/getWriter(...)

java.lang.AssertionError: assertion failed: Shuffle manager not registered for shuffle id: <id>
  at org.apache.spark.shuffle.ShuffleManagerRouter$Cache.remove(ShuffleManagerRouter.scala:122)
  at org.apache.spark.shuffle.ShuffleManagerRouter.unregisterShuffle(ShuffleManagerRouter.scala:64)
```

### Fix direction

Make the router `Cache` tolerate the executor lifecycle: `store` idempotent (`computeIfAbsent`), and `remove`/`unregisterShuffle` handle absence gracefully (return `Option`/`false`) instead of asserting.

--- a/gluten-core/src/main/scala/org/apache/spark/shuffle/ShuffleManagerRouter.scala
+++ b/gluten-core/src/main/scala/org/apache/spark/shuffle/ShuffleManagerRouter.scala
@@ -61,7 +61,10 @@ private class ShuffleManagerRouter(lookup: ShuffleManagerLookup)
   }
 
   override def unregisterShuffle(shuffleId: Int): Boolean = {
-    cache.remove(shuffleId).unregisterShuffle(shuffleId)
+    // On a multi-executor cluster, Spark broadcasts RemoveShuffle to every executor, so a router
+    // that never registered or served this shuffleId still receives unregisterShuffle. Tolerate
+    // the miss instead of asserting: report that this router removed nothing.
+    cache.remove(shuffleId).exists(_.unregisterShuffle(shuffleId))
   }
 
   override def shuffleBlockResolver: ShuffleBlockResolver = resolver
@@ -83,13 +86,12 @@ private class ShuffleManagerRouter(lookup: ShuffleManagerLookup)
           s"${handle.getClass} is not a BaseShuffleHandle so is not supported by " +
             s"GlutenShuffleManager")
     }
-    val shuffleId = baseShuffleHandle.shuffleId
-    if (cache.has(shuffleId)) {
-      return
-    }
-    val dependency = baseShuffleHandle.dependency
-    val manager = lookup.findShuffleManager(dependency)
-    cache.store(shuffleId, manager)
+    // store is idempotent and resolves the manager lazily, so several task threads first-touching
+    // the same new shuffleId install it exactly once, while an already-cached shuffle skips the
+    // lookup entirely.
+    cache.store(
+      baseShuffleHandle.shuffleId,
+      lookup.findShuffleManager(baseShuffleHandle.dependency))
   }
 }
 
@@ -98,17 +100,13 @@ private object ShuffleManagerRouter {
     private val cache: java.util.Map[Int, ShuffleManager] =
       new java.util.concurrent.ConcurrentHashMap()
 
-    def has(shuffleId: Int): Boolean = {
-      cache.containsKey(shuffleId)
-    }
-
-    def store(shuffleId: Int, manager: ShuffleManager): ShuffleManager = {
-      cache.compute(
-        shuffleId,
-        (id, m) => {
-          assert(m == null, s"Shuffle manager was already cached for shuffle id: $id")
-          manager
-        })
+    def store(shuffleId: Int, manager: => ShuffleManager): ShuffleManager = {
+      // Idempotent: on a multi-core executor several task threads may first-touch the same new
+      // shuffleId concurrently. computeIfAbsent lets the first caller install the manager and the
+      // rest observe it, instead of racing into an assertion. The by-name manager is resolved only
+      // on a miss, so an already-cached shuffle skips the lookup. lookup resolves the same manager
+      // for a given dependency, so returning an already-cached entry is safe.
+      cache.computeIfAbsent(shuffleId, _ => manager)
     }
 
     def get(shuffleId: Int): ShuffleManager = {
@@ -117,10 +115,11 @@ private object ShuffleManagerRouter {
       manager
     }
 
-    def remove(shuffleId: Int): ShuffleManager = {
-      val manager = cache.remove(shuffleId)
-      assert(manager != null, s"Shuffle manager not registered for shuffle id: $shuffleId")
-      manager
+    def remove(shuffleId: Int): Option[ShuffleManager] = {
+      // On a multi-executor cluster, Spark broadcasts RemoveShuffle to every executor, so this
+      // router may be asked to remove a shuffleId it never cached. Return None instead of
+      // asserting the entry was present.
+      Option(cache.remove(shuffleId))
     }
 
     def size(): Int = {

--- a/gluten-core/src/test/scala/org/apache/spark/shuffle/ShuffleManagerRouterCacheSuite.scala
+++ b/gluten-core/src/test/scala/org/apache/spark/shuffle/ShuffleManagerRouterCacheSuite.scala
@@ -1,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.shuffle
+
+import org.apache.spark.{ShuffleDependency, SparkConf, TaskContext}
+import org.apache.spark.internal.config.SHUFFLE_MANAGER
+import org.apache.spark.internal.config.UI.UI_ENABLED
+import org.apache.spark.sql.test.SharedSparkSession
+
+import java.util.concurrent.{CopyOnWriteArrayList, CyclicBarrier}
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Reproduces the two ways [[ShuffleManagerRouter]]'s cache mishandles the executor-side lifecycle,
+ * where the router is populated lazily and concurrently by task threads rather than by a single
+ * driver coordinator.
+ */
+class ShuffleManagerRouterCacheSuite extends SharedSparkSession {
+  import ShuffleManagerRouterCacheSuite._
+
+  override protected def sparkConf: SparkConf =
+    super.sparkConf
+      .set(SHUFFLE_MANAGER.key, classOf[GlutenShuffleManager].getName)
+      .set(UI_ENABLED, false)
+
+  override protected def beforeEach(): Unit = {
+    ShuffleManagerRegistry.get().clear()
+    // Register a no-op manager that accepts every dependency, so router lookups resolve
+    // without delegating into a real SortShuffleManager.
+    ShuffleManagerRegistry
+      .get()
+      .register(
+        new LookupKey {
+          override def accepts[K, V, C](dependency: ShuffleDependency[K, V, C]): Boolean = true
+        },
+        classOf[AcceptAllShuffleManager].getName)
+  }
+
+  override protected def afterEach(): Unit = {
+    ShuffleManagerRegistry.get().clear()
+  }
+
+  test("unregisterShuffle tolerates a shuffleId never registered on this router") {
+    // On a multi-executor cluster, Spark broadcasts RemoveShuffle to every executor, so a
+    // router that never registered/served this shuffleId still receives unregisterShuffle.
+    val gm = spark.sparkContext.env.shuffleManager
+    // Force the router to build before exercising it.
+    gm.shuffleBlockResolver
+
+    val neverRegistered = 987654321
+    // Should behave like SortShuffleManager: return a boolean, not throw.
+    assert(!gm.unregisterShuffle(neverRegistered))
+  }
+
+  test("concurrent first-touch of a new shuffleId does not fail tasks") {
+    // On a multi-core executor, N tasks of the same new shuffleId call getReader/getWriter
+    // concurrently. The cache is populated lazily, so they race to register the manager.
+    val gm = spark.sparkContext.env.shuffleManager
+    gm.shuffleBlockResolver // force build
+
+    val concurrency = 4
+    val iterations = 200
+    val idGen = new AtomicInteger(100000)
+    val errors = new CopyOnWriteArrayList[Throwable]()
+
+    (0 until iterations).foreach {
+      _ =>
+        val shuffleId = idGen.getAndIncrement()
+        val barrier = new CyclicBarrier(concurrency)
+        val threads = (0 until concurrency).map {
+          _ =>
+            val t = new Thread(
+              () => {
+                try {
+                  barrier.await()
+                  val dep: ShuffleDependency[Any, Any, Any] = null
+                  val handle = new BaseShuffleHandle(shuffleId, dep)
+                  gm.getReader(
+                    handle,
+                    0,
+                    1,
+                    0,
+                    1,
+                    null.asInstanceOf[TaskContext],
+                    null.asInstanceOf[ShuffleReadMetricsReporter])
+                } catch {
+                  case th: Throwable => errors.add(th)
+                }
+              })
+            t.start()
+            t
+        }
+        threads.foreach(_.join())
+    }
+
+    assert(
+      errors.isEmpty,
+      s"Expected no errors from concurrent first-touch but got ${errors.size}: " +
+        errors.toArray.map(_.toString).toSet.mkString("; "))
+  }
+}
+
+object ShuffleManagerRouterCacheSuite {
+
+  /** A minimal [[ShuffleManager]] that accepts everything and does no real work. */
+  class AcceptAllShuffleManager(conf: SparkConf) extends ShuffleManager {
+    override def registerShuffle[K, V, C](
+        shuffleId: Int,
+        dependency: ShuffleDependency[K, V, C]): ShuffleHandle = {
+      new BaseShuffleHandle(shuffleId, dependency)
+    }
+
+    override def getWriter[K, V](
+        handle: ShuffleHandle,
+        mapId: Long,
+        context: TaskContext,
+        metrics: ShuffleWriteMetricsReporter): ShuffleWriter[K, V] = {
+      null.asInstanceOf[ShuffleWriter[K, V]]
+    }
+
+    override def getReader[K, C](
+        handle: ShuffleHandle,
+        startMapIndex: Int,
+        endMapIndex: Int,
+        startPartition: Int,
+        endPartition: Int,
+        context: TaskContext,
+        metrics: ShuffleReadMetricsReporter): ShuffleReader[K, C] = {
+      null.asInstanceOf[ShuffleReader[K, C]]
+    }
+
+    override def unregisterShuffle(shuffleId: Int): Boolean = true
+
+    override def shuffleBlockResolver: ShuffleBlockResolver = {
+      null.asInstanceOf[ShuffleBlockResolver]
+    }
+
+    override def stop(): Unit = {}
+  }
+}

